Template/update

anoronh4 opened the pull request:
<!--
# mskcc/forte pull request

Many thanks for contributing to mskcc/forte!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the develop branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/mskcc/forte/tree/master/.github/CONTRIBUTING.md)
-->

## Pipeline template update

Some well overdue pipeline template updates. Because there hasn't been a sync in so long i had to follow the [sync retrospectively guide](https://nf-co.re/docs/tutorials/sync/sync_retrospectively), but now that we are in sync we should in the future be able to follow the [manual sync guide](https://nf-co.re/docs/tutorials/sync/manual_sync), provided we don't wait too long between updates. 

Some highlights in the update:
1. pipeline initialization has been structured differently as a subworkflow rather than groovy library methods. prints version info, summarizes and dumps pipeline parameters, validates parameters, and now reads in and validates the input samplesheet.
2. The fastq samplesheet is now validated using nf-schema and the definition file `assets/schema_input.json`. The maf input file is also now validated using nf-schema using a separate definition file `assets/schema_maf_input.json`.
3. software dump prints to a different file in the same location and pipeline template now has different logic for generating it.
5. pipeline completion workflow added, which sends emails upon completion or failure. 

I am also switching the main development branch to `dev` so as to more closely reflect the pipeline and avoid discordance with future template syncs. All github actions triggering with `develop` should now trigger with `dev`, and the base branch of PRs should be `dev`, unless the PR for `dev` to be merged into `main`

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/mskcc/forte/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, also make a PR on the mskcc/forte _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

anoronh4 committed:
update .nf-core.yml

anoronh4 committed:
Template update for nf-core/tools version 2.14.1

anoronh4 committed:
Merge branch 'TEMPLATE' into template/update

anoronh4 committed:
Merge branch 'develop' into template/update

anoronh4 committed:
fix .nf-core.yml linting error

anoronh4 committed:
fix pipeline schema draft

anoronh4 committed:
fix linting errors

anoronh4 committed:
fix linting errors with prettier

anoronh4 committed:
fix uninitialized ch_versions channel

anoronh4 committed:
fix reference to module file

anoronh4 committed:
fix call to getGenomeAttribute method

anoronh4 committed:
Template update for nf-core/tools version 3.1.1

anoronh4 committed:
Merge branch 'TEMPLATE' into template/update

anoronh4 committed:
fix prettier linting issue

anoronh4 committed:
removed redundant parameter in nextflow schema

anoronh4 committed:
fix tests

anoronh4 committed:
update MULTIQC process calls

anoronh4 committed:
fix bugs in meta, replacing 'sample', 'fq_num', 'fastq_pair_id', etc in meta map

anoronh4 committed:
empty

anoronh4 committed:
update CHANGELOG.md for current PR

anoronh4 committed:
add back software yaml file check in the pytest with different file name

anoronh4 committed:
update input schema

anoronh4 committed:
add schema input and validation for maf input sheet

anoronh4 committed:
update fillout test workflow

anoronh4 committed:
disable nf-core tools lint in github actions workflows

anoronh4 committed:
empty

anoronh4 committed:
fix linting errors in input schema

anoronh4 committed:
fixed fillout test with added nf-schema validation

anoronh4 commented:
it might make sense to create a new branch off of `develop` called `dev` and use that instead. the template wants `dev` and our code keeps getting incongruous because i for some reason chose `develop`.

anoronh4 committed:
change branch name develop to dev

anoronh4 committed:
Merge branch 'dev' into template/update

anoronh4 committed:
fix indentations

anoronh4 committed:
Merge branch 'develop' into template/update

anoronh4 committed:
bugfix uncollected multiqc files for batch qc

anoronh4 committed:
bugfix

anoronh4 committed:
update actions/cache to v4

johnoooh commented on CHANGELOG.md:
```diff
@@ -26,6 +27,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - [#132](https://github.com/mskcc/forte/pull/132) - fix generate cff split/span logic for fusioncatcher and arriba
 
+- [#133](https://github.com/mskcc/forte/pull/133) - Template update for nf-core/tools v3.1.1
```
This is the only change in this PR?

anoronh4 replied:
i elaborated a bit more. since the PR is linked someone can always follow the link to get more information.

johnoooh commented on nextflow.config:
```diff
@@ -255,15 +258,43 @@ dag {
+        // TODO nf-core: Update the field with the details of the contributors to your pipeline. New with Nextflow version 24.10.0
+        [
+            name: 'Anne Marie Noronha',
+            affiliation: '',
```
Maybe add your info?

johnoooh commented on subworkflows/local/group_reads.nf:
```diff
@@ -11,7 +11,7 @@ workflow GROUP_READS {
             def meta_clone = meta.clone().findAll { !["read_group","fastq_pair_id"].contains(it.key) }
             meta_clone.id = meta.sample
-            [groupKey(meta_clone,meta.fq_num), reads, read_group, fastq_pair_id]
+            [meta_clone, reads, read_group, fastq_pair_id]
```
Don't need fq_num anymore?  is it in the meta_clone item now?

anoronh4 replied:
i think i was struggling to find a place for fq_num because the input sheet workflow changed so much. either way I don't think it's necessary because without groupKey it should still group the correct number of fastqs anyways? It would make sense to add back if we were using something like watchPath, i believe.

johnoooh reviewed (approved):
Looks good.  Template updates are hard because there are so many little changes spread across so many files but ultimately they shouldnt change too much.  I left a few comments, but they're not urgent.  Overall everything looks good!

anoronh4 committed:
extend the description in the CHANGELOG

anoronh4 committed:
Adding contributor details in nextflow config

pintoa1-mskcc commented on bin/check_samplesheet.py:
why was this deleted?

anoronh4 replied:
the function of this script has been replaced by `subworkflows/local/utils_nfcore_forte_pipeline/main.nf`. the samplesheet is now validated against a json schema file.

pintoa1-mskcc commented on docs/usage.md:
```diff
@@ -208,6 +204,8 @@ If `-profile` is not specified, the pipeline will run locally and expect all sof
 - `apptainer`
   - A generic configuration profile to be used with [Apptainer](https://apptainer.org/)
+- `wave`
+  - A generic configuration profile to enable [Wave](https://seqera.io/wave/) containers. Use together with one of the above (requires Nextflow ` 24.03.0-edge` or later).
```
I think in the readme we should explicitly state that the version of nextflow must be  24.03.0 or later.

anoronh4 replied:
use of wave containerization is not common usage in MSK. currently the pipeline will fail if the version >=24.04.2 because of the how `manifest.nextflowVersion` is set in the config. i think this should be enough information for a user to correct the version of nextflow that they are using if it is too old.

anoronh4 replied:
i did actually see that the readme on the github page has a little badge on it that says 24.04.2 so we have that as well.

pintoa1-mskcc replied:
I meant it'd be good to change it at the readme that shows up when you first enter forte. Right now it says nextflow (>=21.10.3), however with this update it needs to be 24.04.2, as this doesnt just effect this wave container. I didnt specify the container type wave and it still failed when i used 22.01.3

anoronh4 replied:
Sorry if i'm missing it, where does it say nextflow (>=21.10.3)? i see it in the base branch but not in template/update

pintoa1-mskcc replied:
Your totally right I was looking at the wrong branch. ignore me

pintoa1-mskcc commented:
after I updated nextflow to work with this branch, something is wrong with the baits file now?

error I'm getting

> * --baits ({"idt_v2":{"baits":"/juno/work/ccs/cmopipeline/forte/GRCh38_probes/xgen-exome-hyb-panel-v2-probes-hg38.bed","targets":"/juno/work/ccs/cmopipeline/forte/GRCh38_probes/xgen-exome-hyb-panel-v2-targets-hg38.bed"}}): Value is [object] but should be [string]


Curious if its something to do with the change ifrom having params in getGenomeAttribute to removing it in this: https://github.com/mskcc/forte/blob/db0c18fb533482c6a42bacb9f75350477f6d306d/main.nf#L22

pintoa1-mskcc reviewed (dismissed):
I think the new template doesnt work with our assets? Idk something wrong and I cannot run the pipeline on terra

anoronh4 committed:
changed type for bait parameter in the schema file

anoronh4 commented:
> after I updated nextflow to work with this branch, something is wrong with the baits file now?
> 
> error I'm getting
> 
> > * --baits ({"idt_v2":{"baits":"/juno/work/ccs/cmopipeline/forte/GRCh38_probes/xgen-exome-hyb-panel-v2-probes-hg38.bed","targets":"/juno/work/ccs/cmopipeline/forte/GRCh38_probes/xgen-exome-hyb-panel-v2-targets-hg38.bed"}}): Value is [object] but should be [string]
> 
> Curious if its something to do with the change ifrom having params in getGenomeAttribute to removing it in this:
> 
> https://github.com/mskcc/forte/blob/db0c18fb533482c6a42bacb9f75350477f6d306d/main.nf#L22

this error is caused by incorrect configuration in `nextflow_schema.json`. nf-schema was looking through the configs and expecting params.baits to be a `"string"` and not an `"object"`. we definitely want the latter so i've updated that, seems to be working.

anoronh4 commented:
> error I'm getting
> 
> >

is this related to the error with the baits, or something else?

anoronh4 commented:
> I think the new template doesnt work with our assets? Idk something wrong and I cannot run the pipeline on terra

is this related to the error with the baits that you pointed out in the comment just above (and which i believe is now fixed), or something else? or are you still getting the same error?

pintoa1-mskcc commented:
related to my previous comment! thanks for updating
